### DISCUSSION TITLE
Typer improperly parsing args from command defined by `.callback()`

### DISCUSSION THREAD

**Discussion by macintacos** [fastapi/typer#637] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Minimal code available here: https://github.com/macintacos/repro-typer-improper-args
```

### Description

The issue is self-contained in the README here: https://github.com/macintacos/repro-typer-improper-args. Copy/pasting below.

---

[Install `Poetry`](https://python-poetry.org/docs/). Clone the repo and run `poetry install && poetry shell`. Then notice:

- `repro new -o okay lol` successfully runs:

    ```bash
    $ repro new -o okay lol
    args = ['lol']
    opt = 'okay'
    ```

- ...but `repro new okay -o lol` does not:

    ```bash
    $ repro new okay -o lol
    Usage: repro new [OPTIONS] ARGS... COMMAND [ARGS]...
    Try 'repro new -h' for help.
    ╭─ Error ───────────────────────────────────────────╮
    │ Missing option '-o' / '--opt'.                    │
    ╰───────────────────────────────────────────────────╯
    ```

- Notice that the other command, _not_ added as a callback, seems to work okay (the command parameters are identical):

    ```bash
    $ repro actually works lol -o okay
    args = ['lol']
    opt = 'okay'

    $ repro actually works -o okay lol
    args = ['lol']
    opt = 'okay'
    ```

---

For a number of (kinda uninteresting) reasons, I currently have a CLI tool built with the pattern shown in `repro/main.py` and `repo/new.py`.

Can someone confirm that this is a bug? Arguments and Options seem to be improperly being parsed when I do this - when I print things out, it seems like everything is going into the `list[str]`, _including_ the `-o` option and the option's argument as well.

### Operating System

macOS

### Typer Version

0.9.0

### Python Version

Python 3.10.5

**Comment by DiegoAz** [comment]
I don't know if it has anything to do with it but, if I understood correctly, the [documentation](https://typer.tiangolo.com/tutorial/options/name/) about the option names says that you should put the long name first ("--opt", "-o") instead of ("-o", "--opt"). I also think you might want to use `raise typer.Exit()` and `isEager`, both explained [here](https://typer.tiangolo.com/tutorial/options/version/)

**Comment by st3w4r** [comment]
I have the same problem. When I use a **callback decorator**, the flags/options has to come before the argument. It doesn't work if it's after. 

**Does not work:**
```bash
$ cli show ARG --json 
```

**Works:**
```bash
$ cli show --json ARG
```

This makes the experience of the CLI not very user friendly.
